rs/replicated_state/src/replicated_state.rs

```diff
@@ -957,6 +969,94 @@ impl ReplicatedState {
+        let mut priority_queue: BTreeSet<_> = self
+            .canister_states
+            .iter()
+            .filter_map(|(canister_id, canister)| {
+                let memory_usage = canister.system_state.best_effort_message_memory_usage();
+                if memory_usage > ZERO_BYTES {
+                    Some((memory_usage, *canister_id))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        let subnet_queues_memory_usage = self.subnet_queues.best_effort_message_memory_usage();
+        if subnet_queues_memory_usage > 0 {
+            priority_queue.insert((
+                (subnet_queues_memory_usage as u64).into(),
+                self.metadata.own_subnet_id.into(),
+            ));
+        }
+
+        let mut shed_messages = 0;
+        let mut shed_message_bytes = 0.into();
+
+        // Shed messages from the canisters with the largest memory usage until we are
+        // below the limit.
+        //
+        // The `is_empty()` check is a safety net, in case a canister somehow reports
+        // non-zero best-effort memory usage but then fails to shed a message.
+        while memory_usage > limit && !priority_queue.is_empty() {
+            let (memory_usage_before, canister_id) = priority_queue.pop_last().unwrap();
+
+            let message_shed;
+            let memory_usage_after;
+            if canister_id.get() == self.metadata.own_subnet_id.get() {
+                // Shed from the subnet queues.
+                message_shed = self
+                    .subnet_queues
+                    .shed_largest_message(&canister_id, &self.canister_states);
+                memory_usage_after =
+                    (self.subnet_queues.best_effort_message_memory_usage() as u64).into();
+            } else {
+                // Shed from a canister's queues: remove the canister, shed its largest message,
+                // replace it.
+                let mut canister = self.canister_states.remove(&canister_id).unwrap();
+                message_shed = canister
+                    .system_state
+                    .shed_largest_message(&canister_id, &self.canister_states);
+                memory_usage_after = canister.system_state.best_effort_message_memory_usage();
+                self.canister_states.insert(canister_id, canister);
+            }
+            debug_assert!(message_shed);
+
+            // Replace the canister in the priority queue iff its memory usage is still
+            // non-zero AND a message was actually shed.
+            if memory_usage_after > ZERO_BYTES && message_shed {
+                priority_queue.insert((memory_usage_after, canister_id));
+            }
+
+            // Update the total memory usage.
+            let memory_usage_delta = memory_usage_before - memory_usage_after;
+            debug_assert!(memory_usage_delta > ZERO_BYTES);
+            memory_usage -= memory_usage_delta;
```

Comment:
What happens here if `memory_usage_after > memory_usage_before`? I know it shouldn't be possible but and if it were the accounting would be screwed up anyway. But I'd replace this `debug_assert!()` with `debug_assert!(memory_usage_before > memory_usage_after)` and put that in front.

I guess unless `ZERO_BYTES` is intended to be possibly not actually zero bytes for some reason in the future.

Reply:
Good point. Unless `memory_usage_after` was equal to `memory_usage_before`, the `debug_assert!()` wouldn't do anything useful as is. Because what would happen is that the two subtractions would underflow. (OTOH I believe that in a debug build an integer overflow will panic anyway.)

Still, there is one case where shedding a message may result in (marginally) higher memory usage: when we [shed an outbound request](https://github.com/dfinity/ic/blob/8801558493ade5e37a00dfb8ef11f57ecc499415/rs/replicated_state/src/canister_state/queues.rs#L1511-L1522), we enqueue an actual reject response; if the request was really tiny, the reject response may be larger than it (even though the payload only consists of the error message `"Request timed out."`). It is not clear to me what is the right thing to do at that point (apart from not enqueuing an actual response). I suppose we should keep going regardless, because the alternative is to end up with arbitrarily many tiny messages that are never shed.

So I've changed the `debug_assert!()` as suggested, but left everything else as it was.